File: csharp/grade school/GradeSchool.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace GradeSchool
{
    class School
    {
        public IDictionary<int, List<string>> Roster { get; set; }

        public School()
        {
            this.Roster = new Dictionary<int, List<string>>();
        }

        internal void Add(string name, int grade)
        {
            if (!this.Roster.ContainsKey(grade))
            {
                this.Roster.Add(grade, new List<string>());
            }

            if (!this.Roster[grade].Contains(name))
            {
                this.Roster[grade].Add(name);
            }

            this.Roster[grade].Sort();
        }

        internal List<string> Grade(int grade)
        {
            if (!this.Roster.ContainsKey(grade))
            {
                this.Roster.Add(grade, new List<string>());
            }

            return this.Roster[grade];
        }
    }


}

File: fsharp/grade-school/solution/GradeSchool.fs
module GradeSchool

let empty = []

let add name grade school = school @ [(name, grade)]

let grade number school = 
  List.filter (fun student -> snd student = number)  school
  |> List.map (fun x -> fst x)
  |> List.sort

let roster school = 
  List.map (fun student -> snd student) school 
  |> List.distinct
  |> List.map (fun x -> x, (grade x school))
  |> List.sort